### PR TITLE
Consolidate factory API

### PR DESCRIPTION
**Pull request by mavam**
There are numerous factories in VAST who all have the same pattern. This PR consolidates them under a single interface.

- [x] Devise new factory API
- [x] Port `table_slice` factory
- [x] Port `table_slice_builder` factory
- [x] Port `synopsis` factory
- [x] Port `value_index` factory

### CODE REVIEW

File: libvast/vast/value_index.hpp [hunk 1]
```diff
@@ -17,36 +17,42 @@
...
 #include "vast/type.hpp"
+#include "vast/value_index_factory.hpp"
 #include "vast/view.hpp"
 
-#include "vast/detail/assert.hpp"
-#include "vast/detail/overload.hpp"
-
 namespace vast {
 
+using value_index_ptr = caf::intrusive_ptr<value_index>;
```
Comment (Neverlord) [inline libvast/vast/value_index.hpp]: Can we still get away with `unique_ptr` instead?

File: libvast/test/factory.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  concrete(int x, int y) : abstract{x, y} {
+    // nop
+  }
+
+  int f() const override {
+    return x_ * y_;
+  }
+};
+
+std::unique_ptr<abstract> double_make(int x, int y) {
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/test/factory.cpp]: It's actually a somewhat hidden reference to George Orwell. The deliberate ambiguity was on purpose. I thought I'd get away with it in the unit tests.

**Comment by mavam** [comment]
I've addressed almost all feedback. Maybe you let me get away with the slightly ambiguous naming in the unit tests. 😉

**Comment by Neverlord** [comment]
Hm... but just for once. 😉

**Pull request merged**